compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical/dedup_solver/solver.rs

```diff
@@ -0,0 +1,475 @@
+// Quick terminology mapping:
+//     * "rule" => region constraint ("rule" is just shorter)
+//     * "category"/"cat" => group of constraints with similar structure, i.e.
+//                           constraints can only be merged if they're in the
+//                           same category
+//     * "mapping" => mapping from one set of variables to another. Mappings MUST
+//                    be many-to-one. Obviously, they can't be one-to-many, and
+//                    a one-to-one mapping isn't really productive, because
+//                    we're not eliminating any variables that way
+//
+// This algorithm, in its present state, *is* exponential time. This exponential time
+// case happens iff we have potential mappings that overlap, thus requiring them
+// to be combined. For example, if we have a category A that contains constraints involving
+// variables [1, 2, 3] and [11, 12, 4], and a category B that contains constraints involving
+// variables [1, 2, 100] and [11, 12, 101], then the mapping in category A from the first
+// constraint to the second constraint is valid if and only if the mapping from the first
+// constraint to the second constraint in category B is valid (i.e. they depend on each other).
+// In this trivial case, it's obvious that they're both valid, but more complicated cases
+// can create a graph of dependencies (potentially with cycles), creating exponential behavior.
+//
+// In general, I don't think the exponential case should be happening *that* often, and if it does,
+// the dependencies graph shouldn't be very deep, so it shouldn't be terrible. However, I'm not very
+// knowledgable on the types of region constraints that can be generated, so maybe this assertion is false.
+// There's some heuristics that I can use to speed the exponential part up, or maybe just cap the search depth.
+
+use rustc_data_structures::fx::{FxIndexMap, FxIndexSet};
+use std::cell::RefCell;
+use std::collections::BTreeMap;
+
+#[derive(Debug)]
+pub struct DedupSolver {
+    /// The variables present in each rule - the inner vec contains the variables, in the order
+    /// that they appear in the rule
+    rule_vars: Vec<Vec<usize>>,
+    /// The categories that rules can be partitioned into - the inner vec contains all the rules
+    /// that are in the same category
+    rule_cats: Vec<Vec<usize>>,
```

Comment:
instead of using `usize` everywhere, it might be clearer to use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_index/struct.IndexVec.html with a newtype index here.